prov/shm/src/smr_ep.c

```diff
@@ -0,0 +1,747 @@
+	struct smr_cmd *cmd;
+	int peer_id;
+	ssize_t ret = 0;
+	size_t total_len;
+
+	if (msg->iov_count > SMR_IOV_LIMIT) {
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+			"Too many iovs\n");
+		return -FI_EINVAL;
+	}
+
+	ep = container_of(ep_fid, struct smr_ep, util_ep.ep_fid.fid);
+	peer_id = *(int *)ofi_av_get_addr(ep->util_ep.av, msg->addr);
+	fastlock_acquire(&ep->util_ep.tx_cq->cq_lock);
+	if (ofi_cirque_isfull(ep->util_ep.tx_cq->cirq)) {
+		ret = -FI_EAGAIN;
+		goto out;
+	}
+
+	peer_smr = smr_peer_region(ep->region, peer_id);
+	if (ep->region->map->peer_addr[peer_id].name[0] != '\0') {
+		ret = smr_map_to_region(&smr_prov, &peer_smr,
+					ep->region->map->peer_addr[peer_id].name);
+		if (ret) {
+			if (ret == -ENOENT)
+				ret = -FI_EAGAIN; 
+			goto out;
+		}
+	}
+
+	fastlock_acquire(&peer_smr->lock);
+	if (ofi_cirque_isfull(smr_cmd_queue(peer_smr))) {
+		ret = -FI_EAGAIN;
+		goto unlock;
+	}
+
+	total_len = ofi_total_iov_len(msg->msg_iov, msg->iov_count);
+
+	cmd = ofi_cirque_tail(smr_cmd_queue(peer_smr));
+
+	if (total_len <= SMR_CMD_DATA_LEN) {
+		smr_format_send(cmd, msg);
+	} else if (total_len <= SMR_INJECT_SIZE) {
+		freestack_pop(smr_inject_pool(peer_smr), tx_buf);
+		smr_format_inject(cmd, msg, peer_smr, tx_buf);
+	} else {
+		if (ofi_cirque_isfull(smr_resp_queue(ep->region))) {
+			ret = -FI_EAGAIN;
+			goto unlock;
+		}
+		resp = ofi_cirque_tail(smr_resp_queue(ep->region));
+		smr_format_cma(cmd, msg, total_len, ep->region, resp);
+		ofi_cirque_commit(smr_resp_queue(ep->region));
+	}
+	cmd->hdr.cmd_id = (ep->cmd_id)++;
+
+	ofi_cirque_commit(smr_cmd_queue(peer_smr));
+
+	if (total_len <= SMR_INJECT_SIZE && !(flags & SMR_NO_COMPLETION))
+		ep->tx_comp(ep, msg->context);
+
+unlock:
+	fastlock_release(&peer_smr->lock);
+out:
+	fastlock_release(&ep->util_ep.tx_cq->cq_lock);
+	return ret;
+}
+
+ssize_t smr_send(struct fid_ep *ep_fid, const void *buf, size_t len, void *desc,
+		fi_addr_t dest_addr, void *context)
+{
+	struct fi_msg msg;
+	struct iovec msg_iov;
+
+	msg_iov.iov_base = (void *) buf;
+	msg_iov.iov_len = len;
+	msg.msg_iov = &msg_iov;
+	msg.iov_count = 1;
+	msg.addr = dest_addr;
+	msg.context = context;
+
+	return smr_sendmsg(ep_fid, &msg, 0);
+}
+
+
+ssize_t smr_sendv(struct fid_ep *ep_fid, const struct iovec *iov, void **desc,
+		size_t count, fi_addr_t dest_addr, void *context)
+{
+	struct fi_msg msg;
+
+	msg.msg_iov = iov;
+	msg.iov_count = count;
+	msg.addr = dest_addr;
+	msg.context = context;
+
+	return smr_sendmsg(ep_fid, &msg, 0);
+}
+
+ssize_t smr_inject(struct fid_ep *ep_fid, const void *buf, size_t len,
+		fi_addr_t dest_addr)
+{
+	struct fi_msg msg;
+	struct iovec msg_iov;
+
+	if (len > SMR_INJECT_SIZE) {
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+			"message too large to inject\n");
+		return -FI_EINVAL;
+	}
+	msg_iov.iov_base = (void *) buf;
+	msg_iov.iov_len = len;
+	msg.msg_iov = &msg_iov;
+	msg.iov_count = 1;
+	msg.addr = dest_addr;
+
+	return smr_sendmsg(ep_fid, &msg, SMR_NO_COMPLETION | FI_INJECT);
```

Comment:
Rather than calling sendmsg, we should duplicate the code here and optimize out the unneeded checks.